identity-iota/src/tangle/message_version.rs

```diff
@@ -0,0 +1,33 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::Error;
+
+#[derive(Copy, Clone)]
+pub(crate) enum MessageVersion {
+  V1 = 1,
+}
+const CURRENT_MESSAGE_VERSION: MessageVersion = MessageVersion::V1;
+
+/// Adds the current message version flag at the beginning of arbitrary data.
+pub(crate) fn add_version_flag(mut data: Vec<u8>) -> Vec<u8> {
+  let version_flag = CURRENT_MESSAGE_VERSION as u8;
+  data.splice(0..0, [version_flag].iter().cloned());
+  data
+}
+
+/// Checks if flag matches message_version.
+pub(crate) fn check_version_flag(flag: &u8, message_version: MessageVersion) -> Result<(), Error> {
+  if message_version as u8 == *flag {
+    Ok(())
+  } else {
+    Err(Error::InvalidMessageFlags)
+  }
+}
```

Comment:
nit: my personal preference is to declare associated functions rather than naked functions.

E.g.
```rust
impl MessageVersion {
  /// Adds the current message version flag at the beginning of arbitrary data.
  pub(crate) fn add_version_flag(mut data: Vec<u8>) -> Vec<u8> {
    let version_flag = CURRENT_MESSAGE_VERSION as u8;
    data.splice(0..0, [version_flag].iter().cloned());
    data
  }
  
  /// Checks if flag matches message_version.
  pub(crate) fn check_version_flag(flag: &u8, message_version: MessageVersion) -> Result<(), Error> {
    if message_version as u8 == *flag {
      Ok(())
    } else {
      Err(Error::InvalidMessageFlags)
    }
  }
}
```

Reply:
In my understanding associated functions are mostly used as constructors (in Rust) and it feels more idiomatic the way things are now, but I would be happy to be proven wrong. Granted moving the `CURRENT_MESSAGE_VERSION` constant into `MessageVersion` and doing something like this 
```
impl MessageVersion {
  pub(crate) const CURRENT_MESSAGE_VERSION: Self = MessageVersion::V1;
  
/// Adds the current message version flag at the beginning of arbitrary data.
  pub(crate) fn add_version_flag(mut data: Vec<u8>) -> Vec<u8> {
    let version_flag = Self::CURRENT_MESSAGE_VERSION as u8;
    data.splice(0..0, [version_flag].iter().cloned());
    data
  }
  ```
  is maybe neater than how it is now, but I don't see why it should be `MessageVersion`'s "responsibility" to check a version flag, i.e. it feels strange to have `check_version_flag`as an associated function, in which case I would rather have both of them as plain functions defined in this module the way it is now. 

Note that this is mega bike shedding, and I will not complain if @abdulmth refactors it the way you suggested.

Reply:
Update. Having thought about this a bit more I think the following is the "correct" version:

```rust
#[derive(Copy, Clone)]
pub(crate) enum MessageVersion {
  V1 = 1,
}

impl MessageVersion {
  pub(crate) const CURRENT_MESSAGE_VERSION: Self = MessageVersion::V1;
  
/// Adds the current message version flag at the beginning of arbitrary data.
  pub(crate) fn add_version_flag(mut data: Vec<u8>) -> Vec<u8> {
    let version_flag = Self::CURRENT_MESSAGE_VERSION as u8;
    data.splice(0..0, [version_flag].iter().cloned());
    data
  }

  /// Checks if flag matches message_version.
  pub(crate) fn check_version_flag(&self, flag: &u8) -> Result<(), Error> {
    if *self as u8 == *flag {
      Ok(())
    } else {
      Err(Error::InvalidMessageFlags)
    }
  }
}
```

Reply:
Update to the update. Having thought about it even more, it is "correct" the way it is now, except that the 
`add_version_flag` function should probably be renamed to `add_current_version_flag`, and the `CURRENT_MESSAGE_VERSION` constant should be an associated constant `CURRENT` of `MessageVersion`. So all in all it would be :
```rust
#[derive(Copy, Clone)]
pub(crate) enum MessageVersion {
  V1 = 1,
}

impl MessageVersion {
  const CURRENT: Self = MessageVersion::V1; 
}

/// Adds the current message version flag at the beginning of arbitrary data.
pub(crate) fn add_current_version_flag(mut data: Vec<u8>) -> Vec<u8> {
  let version_flag = MessageVersion::CURRENT as u8;
  data.splice(0..0, [version_flag].iter().cloned());
  data
}

/// Checks if flag matches message_version.
pub(crate) fn check_version_flag(flag: &u8, message_version: MessageVersion) -> Result<(), Error> {
  if message_version as u8 == *flag {
    Ok(())
  } else {
    Err(Error::InvalidMessageFlags)
  }
}
```